lib/ansible/modules/windows/win_dhcp_lease.ps1

```diff
@@ -0,0 +1,473 @@
+        "aptr" { $dns_regtype = "AandPTR"; break }
+        "a" { $dns_regtype = "A"; break }
+        "noreg" { $dns_regtype = "NoRegistration"; break }
+        default { $dns_regtype = "NoRegistration"; break }
+    }
+}
+
+Try {
+    # Import DHCP Server PS Module
+    Import-Module DhcpServer
+}
+Catch {
+    # Couldn't load the DhcpServer Module
+    $module.FailJson("The DhcpServer module failed to load properly")
+}
+
+# Determine if there is an existing lease
+if ($ip) {
+    $current_lease = Get-DhcpServerv4Scope | Get-DhcpServerv4Lease | Where-Object IPAddress -eq $ip
+}
+
+# MacAddress was specified
+if ($mac) {
+    if ($mac -like "*-*") {
+        $mac_original = $mac
+        $mac = Convert-MacAddress -mac $mac
+    }
+
+    if ($mac -eq $false) {
+        $module.FailJson("The MAC Address is not properly formatted")
+    }
+    else {
+        $current_lease = Get-DhcpServerv4Scope | Get-DhcpServerv4Lease | Where-Object ClientId -eq $mac_original
+    }
+}
+
+# Did we find a lease/reservation
+if ($current_lease) {
+    $current_lease_exists = $true
+    $original_lease = $current_lease
+    $module.Diff.before = Convert-ReturnValue -Object $original_lease
+}
+else {
+    $current_lease_exists = $false
+}
+
+# If we found a lease, is it a reservation?
+if ($current_lease_exists -eq $true -and ($current_lease.AddressState -like "*Reservation*")) {
+    $current_lease_reservation = $true
+}
+else {
+    $current_lease_reservation = $false
+}
+
+# State: Absent
+# Ensure the DHCP Lease/Reservation is not present
+if ($state -eq "absent") {
+
+    # Required: MAC or IP address
+    if ((-not $mac) -and (-not $ip)) {
+        $module.Result.changed = $false
+        $module.FailJson("The ip or mac parameter is required for state=absent")
+    }
+
+    # If the lease exists, we need to destroy it
+    if ($current_lease_reservation -eq $true) {
+
+        # Try to remove reservation
+        Try {
+            $current_lease | Remove-DhcpServerv4Reservation -WhatIf:$check_mode
+            $state_absent_removed = $true
+        }
+        Catch {
+            $state_absent_removed = $false
+        }
+    }
+    else {
+
+        # Try to remove lease
+        Try {
+            $current_lease | Remove-DhcpServerv4Lease -WhatIf:$check_mode
+            $state_absent_removed = $true
+        }
+        Catch {
+            $state_absent_removed = $false
+        }
+    }
+
+    # If the lease doesn't exist, our work here is done
+    if ($current_lease_exists -eq $false) {
+        $module.Result.changed = $false
+        $module.Result.msg = "The lease or reservation doesn't exist."
+        $module.ExitJson()
+    }
+
+    # See if we removed the lease/reservation
+    if ($state_absent_removed) {
+        $module.Result.changed = $true
+        $module.ExitJson()
+    }
+    else {
+        $module.Result.lease = Convert-ReturnValue -Object $current_lease
+        $module.FailJson("Unable to remove lease/reservation")
```

Comment:
You should save the error record from the try/catch and add it as the 2nd argument. For example if you save the error as `$remove_err` this line would look like

```suggestion
        $module.FailJson("Unable to remove lease/reservation: $($remove_err.Exception.Message)", $remove_err)
```

The reason we do this is so the end user can have at least some indication as to what may have gone wrong.

An even better way would be to no have a try/catch on the original calls in the first place. Exception handling should only be used if you can actually recover from particular exceptions, or you need to add further context to the error message which this is doing neither.